httpd/handler.go

```diff
@@ -174,11 +178,94 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, user *influ
 		return
 	}
 
-	// Execute query. One result will return for each statement.
-	results := h.server.ExecuteQuery(query, db, user)
+	// get the chunking settings
+	chunked := q.Get("chunked") == "true"
+	// even if we're not chunking, the engine will chunk at this size and then the handler will combine results
+	chunkSize := DefaultChunkSize
+	if chunked {
+		if cs, err := strconv.ParseInt(q.Get("chunk_size"), 10, 64); err == nil {
+			chunkSize = int(cs)
+		}
+	}
 
 	// Send results to client.
-	httpResults(w, results, pretty)
+	w.Header().Add("content-type", "application/json")
+	results, err := h.server.ExecuteQuery(query, db, user, chunkSize)
+	if err != nil {
+		if isAuthorizationError(err) {
+			w.WriteHeader(http.StatusUnauthorized)
+		} else {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		return
+	}
+
+	// if we're not chunking, this will be the in memory buffer for all results before sending to client
+	res := influxdb.Results{Results: make([]*influxdb.Result, 0)}
+	statusWritten := false
+
+	// pull all results from the channel
+	for r := range results {
+		// write the status header based on the first result returned in the channel
+		if !statusWritten {
+			if r != nil && r.Err != nil {
+				if isAuthorizationError(r.Err) {
+					w.WriteHeader(http.StatusUnauthorized)
+				} else if isMeasurementNotFoundError(r.Err) {
+					w.WriteHeader(http.StatusOK)
+				} else if isFieldNotFoundError(r.Err) {
+					w.WriteHeader(http.StatusOK)
+				} else {
+					w.WriteHeader(http.StatusInternalServerError)
+				}
+			} else {
+				w.WriteHeader(http.StatusOK)
+			}
+			statusWritten = true
+		}
+
+		// ignore nils
+		if r == nil {
+			continue
+		}
+
+		// if chunked we write out this result and flush
+		if chunked {
+			res.Results = []*influxdb.Result{r}
+			w.Write(marshalPretty(res, pretty))
+			w.(http.Flusher).Flush()
+			continue
+		}
+
+		// it's not chunked so buffer results in memory.
+		// results for statements need to be combined together. We need to check if this new result is
+		// for the same statement as the last result, or for the next statement
+		l := len(res.Results)
+		if l == 0 {
+			res.Results = append(res.Results, r)
+		} else if res.Results[l-1].StatementID == r.StatementID {
+			cr := res.Results[l-1]
+			cr.Series = append(cr.Series, r.Series...)
+		} else {
+			res.Results = append(res.Results, r)
+		}
+	}
+
+	// if it's not chunked we buffered everything in memory, so write it out
+	if !chunked {
+		w.Write(marshalPretty(res, pretty))
+	}
+}
+
+// marshalPretty will marshal the interface to json either pretty printed or not
+func marshalPretty(r interface{}, pretty bool) []byte {
+	var b []byte
+	if pretty {
+		b, _ = json.MarshalIndent(r, "", "    ")
```

Comment:
We really shouldn't ignore errors here.  It's likely we won't cause the error, but if we ran out of memory, disk full, etc. this can break, and then we will be chasing a ghost bug as we will think our code is buggy, when in fact there was a legitimate reason the json couldn't marshal.